src/main/kotlin/systems/ajax/malov/stockanalyzer/repository/impl/MongoStockRecordRepository.kt

```diff
@@ -0,0 +1,209 @@
+            }
+    }
+
+    private fun getProjectWithAvgMaxValues(maxChange: BigDecimal?, maxPercentChange: BigDecimal?): Bson {
+        val records = current().getArray<MqlDocument>("records")
+        return project(
+            fields(
+                exclude("_id"),
+                include("records"),
+                computed(
+                    "avgChange",
+                    gradeAverage(records, MongoStockRecord::change.name)
+                ),
+                computed(
+                    "avgPercentChange",
+                    gradeAverage(records, MongoStockRecord::percentChange.name)
+                ),
+                maxChange?.let {
+                    computed(
+                        "maxChange",
+                        Document("\$toDecimal", it.toString())
+                    )
+                } ?: Document("maxChange", null),
+                maxPercentChange?.let {
+                    computed(
+                        "maxPercentChange",
+                        Document("\$toDecimal", it.toString())
+                    )
+                } ?: Document("maxPercentChange", null),
+                computed(
+                    "changeCoef",
+                    Document("\$toDecimal", WEIGHTED_COEFFICIENT_FOR_PRICE_COEFFICIENTS.toString())
+                ),
+                computed(
+                    "percentChangeCoef",
+                    Document("\$toDecimal", WEIGHTED_COEFFICIENT_FOR_PRICE_COEFFICIENTS.toString())
+                ),
+            )
+        )
+    }
+
+    private fun getWeightedPipeLine(): Bson {
+        val firstCoef = current().getNumber("avgChange")
+            .divide(current().getNumber("maxChange"))
+            .multiply(current().getNumber("changeCoef"))
+        val secondCoef = current().getNumber("avgPercentChange")
+            .divide(current().getNumber("maxPercentChange"))
+            .multiply(current().getNumber("percentChangeCoef"))
+
+        val weight = firstCoef.add(secondCoef)
+
+        val projectionWithWeightCoef = project(
+            fields(
+                include("records"),
+                computed(
+                    "weight", weight
+                )
+            )
+        )
+        return projectionWithWeightCoef
+    }
+
+    private fun gradeAverage(students: MqlArray<MqlDocument>, fieldName: String): MqlNumber {
+        return students
+            .sum { student -> student.getInteger(fieldName) }
+            .divide(students.size())
+    }
+
+    private fun getOnlyMostRecentNDataRecords(map: Map<String, MutableList<MongoStockRecord>>)
+            : Map<String, List<MongoStockRecord>> {
+        return map.mapValues {
+            it.value.sortedByDescending { stockRecord -> stockRecord.dateOfRetrieval }
+                .take(NUMBER_OF_HISTORY_RECORDS_PER_STOCK_SYMBOL)
+        }
+    }
+
+    private fun convertResultsToList(results: List<Document>): Map<String, MutableList<MongoStockRecord>> {
```

Comment:
maybe introduce an intermediate model for the result so that the document can me deserialised by spring data?